src/dotnet/CarbonAware.Tools.WattTimeClient.Tests/WattTimeClientTests.cs

```diff
@@ -0,0 +1,545 @@
+
+            this.Options = new Mock<IOptionsMonitor<WattTimeClientConfiguration>>();
+            this.Log = new Mock<ILogger<WattTimeClient.WattTimeClient>>();
+
+            this.Options.Setup(o => o.CurrentValue).Returns(() => this.Configuration);
+
+            this.ActivitySource = new ActivitySource("CarbonAware.Tools.WattTimeClient.Tests");
+
+            this.BasicAuthValue = Convert.ToBase64String(Encoding.ASCII.GetBytes($"{this.Configuration.Username}:{this.Configuration.Password}"));
+        }
+
+        [Test]
+        public void AllPublicMethods_ThrowsWhenInvalidLogin()
+        {
+            this.CreateHttpClient(m =>
+            {
+                var response = this.MockWattTimeAuthResponse(m, new StringContent(""), "token");
+                return Task.FromResult(response);
+            });
+
+            this.BasicAuthValue = "invalid";
+            var client = new WattTimeClient.WattTimeClient(this.HttpClient, this.Options.Object, this.Log.Object, this.ActivitySource);
+            
+            Assert.ThrowsAsync<WattTimeClientException>(async () => await client.GetDataAsync("ba", "start", "end"));
+            Assert.ThrowsAsync<WattTimeClientException>(async () => await client.GetCurrentForecastAsync("ba"));
+            Assert.ThrowsAsync<WattTimeClientException>(async () => await client.GetForecastByDateAsync("ba", "start", "end"));
+            Assert.ThrowsAsync<WattTimeClientException>(async () => await client.GetBalancingAuthorityAsync("lat", "long"));
+            Assert.ThrowsAsync<WattTimeClientException>(async () => await client.GetBalancingAuthorityAbbreviationAsync("lat", "long"));
+            Assert.ThrowsAsync<WattTimeClientException>(async () => await client.GetHistoricalDataAsync("ba"));
+        }
+
+        [Test]
+        public void GetDataAsync_ThrowsWhenBadJsonIsReturned()
+        {
+            this.CreateHttpClient(m =>
+            {
+                var response = this.MockWattTimeAuthResponse(m, new StringContent("This is bad json."));
+                return Task.FromResult(response);
+            });
+
+            var client = new WattTimeClient.WattTimeClient(this.HttpClient, this.Options.Object, this.Log.Object, this.ActivitySource);
+            client.SetBearerAuthenticationHeader(this.DefaultTokenValue);
+
+            Assert.ThrowsAsync<JsonException>(async () => await client.GetDataAsync("ba", "start", "end"));
+        }
+
+
+        [Test]
+        public async Task GetDataAsync_DeserializesExpectedResponse()
+        {
+            this.CreateHttpClient(m =>
+            {
+                Assert.AreEqual("https://api2.watttime.org/v2/data?ba=balauth&starttime=start&endtime=end", m.RequestUri?.ToString());
+                Assert.AreEqual(HttpMethod.Get, m.Method);
+                var response = this.MockWattTimeAuthResponse(m, new StringContent("[" +
+                    "{\"ba\":\"ba\",\"datatype\":\"dt\",\"frequency\": 300,\"market\":\"mkt\"," +
```

Comment:
It feels that this is error prune. Perhaps having a internal class using json lib to create this output. Easy to maintain imo.

Reply:
Fixed